scalyr_agent/log_processing.py

```diff
@@ -3318,6 +3320,8 @@ def find_matches(
             then if copy_at_index_zero is True, the file will be processed from the first byte in the file.  Otherwise,
             the processing will skip over all bytes currently in the file and only process bytes added after this
             point.
+        @param get_log_processor_factory:  Callable which returns another callable which creates
```

Comment:
Yes, I made it more complicated than it has to be.

The new approach is that the LogMatchers do not create log processors directly by instantiating the 'LogFileProcessor' class. Instead of that, we find an appropriate worker pool for this matcher. (by the `api_key_id`) and then pass a new method of the worker pool named `create_and_add_new_log_processor` as a `create_log_processor` parameter.

Inside the `find_matches` method, the LogMatcher will call the `create_log_processor` parameter to create the log processor.

https://github.com/scalyr/scalyr-agent-2/blob/d0dc47790615f14fc7ca209b0f985b02399a668c/scalyr_agent/sharded_copying_manager/copying_manager.py#L1091-L1102


**How `worker_pool.create_and_add_new_log_processor` creates  a new log processor?**

This method finds a worker and calls its method - `create_and_schedule_new_log_processor` 

https://github.com/scalyr/scalyr-agent-2/blob/d0dc47790615f14fc7ca209b0f985b02399a668c/scalyr_agent/sharded_copying_manager/copying_manager.py#L112-L123

This method finally creates the new log processor and also immediately adds it to the worker itself. 

https://github.com/scalyr/scalyr-agent-2/blob/d0dc47790615f14fc7ca209b0f985b02399a668c/scalyr_agent/sharded_copying_manager/worker.py#L1080-L1097

In other words, the log processor creation and adding it to the worker is now combined into one operation.